File: backend/app/services/fuzzy_matcher.py
import math
from dataclasses import dataclass
from datetime import datetime
from decimal import Decimal
from typing import Any, Dict, List, Optional, Tuple
from rapidfuzz import fuzz
from app.core.config import settings
from app.models import Transaction


@dataclass
class FuzzyScoreBreakdown:
    overall_confidence: Decimal
    amount_score: Decimal
    date_score: Decimal
    description_score: Decimal
    reference_score: Decimal
    explanation: str

    def to_dict(self) -> Dict[str, Any]:
        return {
            "overall_confidence": str(self.overall_confidence),
            "amount_score": str(self.amount_score),
            "date_score": str(self.date_score),
            "description_score": str(self.description_score),
            "reference_score": str(self.reference_score),
            "explanation": self.explanation,
        }


class FuzzyMatcher:
    """Intelligent fuzzy matching and explainable confidence scoring engine."""

    @staticmethod
    def compute_amount_score(amount_s: Decimal, amount_l: Decimal) -> Tuple[Decimal, str]:
        """Score monetary amount agreement with linear decay for small fee/rounding variances."""
        if amount_s == amount_l:
            return Decimal("1.0000"), "Amount matches exactly (100%)"

        # If signs differ, they cannot be matched
        if (amount_s > 0 and amount_l < 0) or (amount_s < 0 and amount_l > 0):
            return Decimal("0.0000"), "Amount sign mismatch (0%)"

        abs_s = abs(amount_s)
        abs_l = abs(amount_l)
        max_abs = max(abs_s, abs_l)
        if max_abs == Decimal("0"):
            return Decimal("1.0000"), "Zero amount exact match (100%)"

        diff = abs(abs_s - abs_l)
        pct_diff = float(diff / max_abs)

        # Allow small percentage variance up to 5%
        if pct_diff <= 0.05:
            score = max(0.0, 1.0 - (pct_diff * 4.0))
            score_dec = Decimal(str(round(score, 4)))
            return score_dec, f"Amount variance of ${diff:.2f} ({score_dec * 100:.1f}%)"

        return Decimal("0.0000"), f"Amount variance ${diff:.2f} exceeds tolerance (0%)"

    @staticmethod
    def compute_date_score(date_s: datetime, date_l: datetime) -> Tuple[Decimal, str]:
        """Score transaction date proximity."""
        delta_days = abs((date_s.date() - date_l.date()).days)

        if delta_days == 0:
            return Decimal("1.0000"), "Same transaction date (100%)"
        elif delta_days == 1:
            return Decimal("0.9500"), "Date differs by 1 day (95%)"
        elif delta_days == 2:
            return Decimal("0.9000"), "Date differs by 2 days (90%)"
        elif delta_days == 3:
            return Decimal("0.8000"), "Date differs by 3 days (80%)"
        elif delta_days <= settings.FUZZY_MAX_DATE_DIFF_DAYS:
            score = max(0.40, 1.0 - (delta_days * 0.09))
            score_dec = Decimal(str(round(score, 4)))
            return score_dec, f"Date differs by {delta_days} days ({score_dec * 100:.1f}%)"

        return Decimal("0.0000"), f"Date difference of {delta_days} days exceeds threshold (0%)"

    @staticmethod
    def compute_description_score(desc_s: Optional[str], desc_l: Optional[str]) -> Tuple[Decimal, str]:
        """Score normalized description similarity using RapidFuzz."""
        if not desc_s or not desc_l:
            return Decimal("0.0000"), "Missing description (0%)"

        s_clean = desc_s.strip()
        l_clean = desc_l.strip()

        if s_clean == l_clean:
            return Decimal("1.0000"), "Description identical (100%)"

        # Calculate Token Set Ratio & Levenshtein Ratio
        token_ratio = fuzz.token_set_ratio(s_clean, l_clean)
        partial_ratio = fuzz.partial_ratio(s_clean, l_clean)
        std_ratio = fuzz.ratio(s_clean, l_clean)

        # Weighted combination favoring token containment
        best_sim = max(token_ratio, (partial_ratio * 0.7 + std_ratio * 0.3)) / 100.0
        score_dec = Decimal(str(round(best_sim, 4)))

        return score_dec, f"Description similarity {score_dec * 100:.1f}% ('{desc_s}' ~ '{desc_l}')"

    @staticmethod
    def compute_reference_score(ref_s: Optional[str], ref_l: Optional[str]) -> Tuple[Decimal, Optional[str]]:
        """Score external reference similarity."""
        if not ref_s or not ref_l:
            return Decimal("0.0000"), None

        s_clean = ref_s.strip().upper()
        l_clean = ref_l.strip().upper()

        if s_clean == l_clean:
            return Decimal("1.0000"), "Reference exact match (100%)"

        if s_clean in l_clean or l_clean in s_clean:
            return Decimal("0.9000"), f"Reference partial containment ({s_clean} in {l_clean})"

        sim = fuzz.ratio(s_clean, l_clean) / 100.0
        score_dec = Decimal(str(round(sim, 4)))
        return score_dec, f"Reference similarity {score_dec * 100:.1f}%"

    @staticmethod
    def score_pair(stmt: Transaction, ledger: Transaction) -> FuzzyScoreBreakdown:
        """Compute holistic weighted fuzzy match confidence score and explanation."""
        amount_score, amt_exp = FuzzyMatcher.compute_amount_score(stmt.amount, ledger.amount)
        date_score, date_exp = FuzzyMatcher.compute_date_score(stmt.transaction_date, ledger.transaction_date)
        desc_score, desc_exp = FuzzyMatcher.compute_description_score(
            stmt.normalized_description or stmt.description,
            ledger.normalized_description or ledger.description,
        )
        ref_score, ref_exp = FuzzyMatcher.compute_reference_score(
            stmt.normalized_reference or stmt.external_reference,
            ledger.normalized_reference or ledger.external_reference,
        )

        w_amt = Decimal(str(settings.FUZZY_WEIGHT_AMOUNT))
        w_date = Decimal(str(settings.FUZZY_WEIGHT_DATE))
        w_desc = Decimal(str(settings.FUZZY_WEIGHT_DESCRIPTION))
        w_ref = Decimal(str(settings.FUZZY_WEIGHT_REFERENCE))

        if ref_exp is not None:
            # All 4 features active
            overall = (amount_score * w_amt) + (date_score * w_date) + (desc_score * w_desc) + (ref_score * w_ref)
        else:
            # Rebalance weights over available 3 features (amount, date, description)
            total_w = w_amt + w_date + w_desc
            overall = ((amount_score * w_amt) + (date_score * w_date) + (desc_score * w_desc)) / total_w

        overall_clamped = min(Decimal("1.0000"), max(Decimal("0.0000"), overall.quantize(Decimal("0.0001"))))

        # Build structured explanation sentence
        explanations = [amt_exp, date_exp, desc_exp]
        if ref_exp:
            explanations.append(ref_exp)

        explanation_str = f"Confidence {overall_clamped * 100:.1f}%: {'; '.join(explanations)}."

        return FuzzyScoreBreakdown(
            overall_confidence=overall_clamped,
            amount_score=amount_score,
            date_score=date_score,
            description_score=desc_score,
            reference_score=ref_score,
            explanation=explanation_str,
        )

    @staticmethod
    def generate_candidate_pairs(
        statement_txs: List[Transaction],
        ledger_txs: List[Transaction],
    ) -> List[Tuple[Transaction, Transaction, FuzzyScoreBreakdown]]:
        """
        Generate and score candidate pairs using intelligent search pruning:
        - Date within max date window
        - Amount within max percentage tolerance or identical signs
        """
        candidate_results: List[Tuple[Transaction, Transaction, FuzzyScoreBreakdown]] = []

        for s in statement_txs:
            for l in ledger_txs:
                # 1. Quick Pruning: Signs must match
                if (s.amount > 0 and l.amount < 0) or (s.amount < 0 and l.amount > 0):
                    continue

                # 2. Quick Pruning: Date distance check
                delta_days = abs((s.transaction_date.date() - l.transaction_date.date()).days)
                if delta_days > settings.FUZZY_MAX_DATE_DIFF_DAYS:
                    continue

                # 3. Quick Pruning: Amount percentage check
                max_amt = max(abs(s.amount), abs(l.amount))
                if max_amt > 0:
                    pct = float(abs(abs(s.amount) - abs(l.amount)) / max_amt)
                    if pct > settings.FUZZY_MAX_AMOUNT_DIFF_PERCENT:
                        continue

                # Score candidate pair
                score = FuzzyMatcher.score_pair(s, l)
                if score.overall_confidence > Decimal("0.30"):
                    candidate_results.append((s, l, score))

        # Sort candidate pairs by overall confidence descending
        candidate_results.sort(key=lambda item: item[2].overall_confidence, reverse=True)
        return candidate_results

File: server/src/reconciliation/fuzzyMatcher.ts
import { config } from "../config/index.js";
import type { FuzzyScoreResult } from "../types/index.js";

export class FuzzyMatcher {
  /**
   * Levenshtein Distance similarity ratio (0.0 to 1.0).
   */
  static levenshteinRatio(s1: string, s2: string): number {
    if (s1 === s2) return 1.0;
    if (!s1 || !s2) return 0.0;

    const len1 = s1.length;
    const len2 = s2.length;
    const matrix: number[][] = [];

    for (let i = 0; i <= len1; i++) {
      matrix[i] = [i];
    }
    for (let j = 0; j <= len2; j++) {
      matrix[0][j] = j;
    }

    for (let i = 1; i <= len1; i++) {
      for (let j = 1; j <= len2; j++) {
        const cost = s1[i - 1] === s2[j - 1] ? 0 : 1;
        matrix[i][j] = Math.min(
          matrix[i - 1][j] + 1,
          matrix[i][j - 1] + 1,
          matrix[i - 1][j - 1] + cost,
        );
      }
    }

    const dist = matrix[len1][len2];
    const maxLen = Math.max(len1, len2);
    return maxLen === 0 ? 1.0 : Math.max(0.0, 1.0 - dist / maxLen);
  }

  /**
   * Token Set / Jaccard similarity between words in descriptions.
   */
  static tokenSimilarity(s1: string, s2: string): number {
    const tokens1 = new Set(s1.toLowerCase().split(/\s+/).filter(Boolean));
    const tokens2 = new Set(s2.toLowerCase().split(/\s+/).filter(Boolean));

    if (tokens1.size === 0 && tokens2.size === 0) return 1.0;
    if (tokens1.size === 0 || tokens2.size === 0) return 0.0;

    let intersection = 0;
    tokens1.forEach((t) => {
      if (tokens2.has(t)) intersection++;
    });

    const union = new Set([...tokens1, ...tokens2]).size;
    return union === 0 ? 1.0 : intersection / union;
  }

  /**
   * Compute monetary amount similarity with small variance / fee tolerance.
   */
  static computeAmountScore(amt1: number, amt2: number): { score: number; explanation: string } {
    if (amt1 === amt2) {
      return { score: 1.0, explanation: "Amount matches exactly (100%)" };
    }

    // Opposite signs cannot match
    if ((amt1 > 0 && amt2 < 0) || (amt1 < 0 && amt2 > 0)) {
      return { score: 0.0, explanation: "Amount sign mismatch (0%)" };
    }

    const abs1 = Math.abs(amt1);
    const abs2 = Math.abs(amt2);
    const maxAbs = Math.max(abs1, abs2);
    if (maxAbs === 0) return { score: 1.0, explanation: "Zero amount match (100%)" };

    const diff = Math.abs(abs1 - abs2);
    const pctDiff = diff / maxAbs;

    if (pctDiff <= config.FUZZY_MAX_AMOUNT_DIFF_PERCENT) {
      const score = Math.max(0.0, 1.0 - pctDiff * 4.0);
      return {
        score: Math.round(score * 10000) / 10000,
        explanation: `Amount variance of $${diff.toFixed(2)} (${(score * 100).toFixed(1)}%)`,
      };
    }

    return {
      score: 0.0,
      explanation: `Amount variance of $${diff.toFixed(2)} exceeds tolerance (0%)`,
    };
  }

  /**
   * Compute date proximity score.
   */
  static computeDateScore(d1: Date, d2: Date): { score: number; explanation: string } {
    const diffMs = Math.abs(d1.getTime() - d2.getTime());
    const deltaDays = Math.floor(diffMs / (1000 * 60 * 60 * 24));

    if (deltaDays === 0) {
      return { score: 1.0, explanation: "Same transaction date (100%)" };
    } else if (deltaDays === 1) {
      return { score: 0.95, explanation: "Date differs by 1 day (95%)" };
    } else if (deltaDays === 2) {
      return { score: 0.9, explanation: "Date differs by 2 days (90%)" };
    } else if (deltaDays === 3) {
      return { score: 0.8, explanation: "Date differs by 3 days (80%)" };
    } else if (deltaDays <= config.FUZZY_MAX_DATE_DIFF_DAYS) {
      const score = Math.max(0.4, 1.0 - deltaDays * 0.09);
      return {
        score: Math.round(score * 10000) / 10000,
        explanation: `Date differs by ${deltaDays} days (${(score * 100).toFixed(1)}%)`,
      };
    }

    return {
      score: 0.0,
      explanation: `Date difference of ${deltaDays} days exceeds window (0%)`,
    };
  }

  /**
   * Compute normalized description similarity.
   */
  static computeDescriptionScore(desc1?: string | null, desc2?: string | null): { score: number; explanation: string } {
    if (!desc1 || !desc2) {
      return { score: 0.0, explanation: "Missing description" };
    }

    const d1 = desc1.trim().toLowerCase();
    const d2 = desc2.trim().toLowerCase();

    if (d1 === d2) {
      return { score: 1.0, explanation: "Description identical (100%)" };
    }

    const tokenSim = this.tokenSimilarity(d1, d2);
    const levSim = this.levenshteinRatio(d1, d2);

    // Blended score favoring token containment and substring overlap
    const combined = Math.max(tokenSim, levSim * 0.4 + tokenSim * 0.6);
    const score = Math.round(combined * 10000) / 10000;

    return {
      score,
      explanation: `Description similarity ${(score * 100).toFixed(1)}% ('${desc1}' ~ '${desc2}')`,
    };
  }

  /**
   * Compute external reference similarity.
   */
  static computeReferenceScore(ref1?: string | null, ref2?: string | null): { score: number; explanation: string | null } {
    if (!ref1 || !ref2) {
      return { score: 0.0, explanation: null };
    }

    const r1 = ref1.trim().toUpperCase();
    const r2 = ref2.trim().toUpperCase();

    if (r1 === r2) {
      return { score: 1.0, explanation: "Reference exact match (100%)" };
    }

    if (r1.includes(r2) || r2.includes(r1)) {
      return { score: 0.9, explanation: `Reference partial match (${r1} ~ ${r2})` };
    }

    const sim = this.levenshteinRatio(r1, r2);
    const score = Math.round(sim * 10000) / 10000;
    return { score, explanation: `Reference similarity ${(score * 100).toFixed(1)}%` };
  }

  /**
   * Compute overall explainable confidence score.
   */
  static scorePair(stmtTx: {
    amount: number;
    transactionDate: Date;
    normalizedDescription?: string | null;
    description: string;
    normalizedReference?: string | null;
    externalReference?: string | null;
  }, ledgerTx: {
    amount: number;
    transactionDate: Date;
    normalizedDescription?: string | null;
    description: string;
    normalizedReference?: string | null;
    externalReference?: string | null;
  }): FuzzyScoreResult {
    const amtResult = this.computeAmountScore(stmtTx.amount, ledgerTx.amount);
    const dateResult = this.computeDateScore(stmtTx.transactionDate, ledgerTx.transactionDate);
    const descResult = this.computeDescriptionScore(
      stmtTx.normalizedDescription || stmtTx.description,
      ledgerTx.normalizedDescription || ledgerTx.description,
    );
    const refResult = this.computeReferenceScore(
      stmtTx.normalizedReference || stmtTx.externalReference,
      ledgerTx.normalizedReference || ledgerTx.externalReference,
    );

    const wAmt = config.FUZZY_WEIGHT_AMOUNT;
    const wDesc = config.FUZZY_WEIGHT_DESCRIPTION;
    const wDate = config.FUZZY_WEIGHT_DATE;
    const wRef = config.FUZZY_WEIGHT_REFERENCE;

    let overall = 0;
    if (refResult.explanation !== null) {
      overall = amtResult.score * wAmt + descResult.score * wDesc + dateResult.score * wDate + refResult.score * wRef;
    } else {
      const totalW = wAmt + wDesc + wDate;
      overall = (amtResult.score * wAmt + descResult.score * wDesc + dateResult.score * wDate) / totalW;
    }

    const clamped = Math.min(1.0, Math.max(0.0, Math.round(overall * 10000) / 10000));
    const explanations = [amtResult.explanation, dateResult.explanation, descResult.explanation];
    if (refResult.explanation) explanations.push(refResult.explanation);

    return {
      overallConfidence: clamped,
      amountScore: amtResult.score,
      dateScore: dateResult.score,
      descriptionScore: descResult.score,
      referenceScore: refResult.score,
      explanation: `Confidence ${(clamped * 100).toFixed(1)}%: ${explanations.join("; ")}.`,
    };
  }
}
